src/ServiceProviders/RepositoryProviders/GitLab/GitLabProvider.php

```diff
@@ -0,0 +1,246 @@
+    const GITLAB_URL_DEFAULT = 'gitlab.com';
+    protected $config;
+
+    protected $repositoryEnvironment;
+
+    public function __construct(Config $config)
+    {
+        $this->config = $config;
+        $this->GITLAB_URL = $config->get(self::GITLAB_CONFIG_PATH, self::GITLAB_URL_DEFAULT);
+    }
+
+    public function infer($url)
+    {
+        return strpos($url, $this->GITLAB_URL) !== false;
+    }
+
+    public function getEnvironment()
+    {
+        if (!$this->repositoryEnvironment) {
+            $this->repositoryEnvironment = (new RepositoryEnvironment())
+            ->setServiceName(self::SERVICE_NAME);
+        }
+        return $this->repositoryEnvironment;
+    }
+
+    public function tokenKey()
+    {
+        return self::GITLAB_TOKEN;
+    }
+
+    public function hasToken()
+    {
+        $repositoryEnvironment = $this->getEnvironment();
+        return $repositoryEnvironment->hasToken();
+    }
+
+    public function token()
+    {
+        $repositoryEnvironment = $this->getEnvironment();
+        return $repositoryEnvironment->token();
+    }
+
+    public function setToken($token)
+    {
+        $repositoryEnvironment = $this->getEnvironment();
+        $repositoryEnvironment->setToken($this->tokenKey(), $token);
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function credentialRequests()
+    {
+        // Tell the credential manager that we require one credential: the
+        // GITLAB_TOKEN that will be used to authenticate.
+        $gitlabTokenRequest = new CredentialRequest(
+            $this->tokenKey(),
+            "Please generate a GitLab personal access token by visiting the page:\n\n    https://" . $this->GITLAB_URL . "/profile/personal_access_tokens\n\n For more information, see:\n\n    https://" . $this->GITLAB_URL . "/help/user/profile/personal_access_tokens.md.\n\n Give it the 'api' (required) scopes.",
+            "Enter GitLab personal access token: ",
+            '#^[0-9a-zA-Z\-]{20}$#',
+            'GitLab authentication tokens should be 20-character strings containing only the letters a-z and digits (0-9). Please enter your token again.'
+        );
+
+        return [ $gitlabTokenRequest ];
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function setCredentials(CredentialProviderInterface $credentials_provider)
+    {
+        // Since the `credentialRequests()` method declared that we need a
+        // GITLAB_TOKEN credential, it will be available for us to copy from
+        // the credentials provider when this method is called.
+        $tokenKey = $this->tokenKey();
+        $token = $credentials_provider->fetch($tokenKey);
+        if (!$token) {
+            throw new \Exception('Could not determine authentication token for GitLab serivces. Please set ' . $tokenKey);
+        }
+        $this->setToken($token);
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function authenticatedUser()
+    {
+        $userData = $this->gitLabAPI('user');
```

Comment:
During testing I noticed your Gitlab provider doesn't work without an `--org` parameter and while debugging I found out that the API URL in this line is not correct, please replace it with this:

`$userData = $this->gitLabAPI('api/v4/user');`